### PR TITLE
feat: Add versioning support for FAISS online store

### LINKED ISSUE #6173: Add feature view versioning support to Faiss online store
## Summary

Add versioned feature view read/write support to the `Faiss` online store, so that version-qualified feature references (e.g., `driver_stats@v2:trips_today`) resolve to the correct versioned data.

## Context

Feature view versioning was introduced in #6101. Currently, only the SQLite online store supports versioned reads/writes. All other online stores raise `VersionedOnlineReadNotSupported` when a versioned ref is used.

The implementation pattern (see `sqlite.py` for reference) involves:
1. **Write path**: When `enable_online_feature_view_versioning` is enabled, materialize data into a version-specific table/key namespace (e.g., `{project}_{fv_name}_v{version}`)
2. **Read path**: Route version-qualified lookups to the correct versioned table/key namespace

## Reference implementation

- SQLite online store: `sdk/python/feast/infra/online_stores/sqlite.py`
- Store file: `sdk/python/feast/infra/online_stores/faiss_online_store.py`

## Related

Part of #2728

### PR DESCRIPTION
**Pull request by rpathade**
# What this PR does / why we need it:
Closes #6173

Feature view versioning was introduced in #6101, but version-qualified feature references (e.g. `driver_stats@v2:trips_today`) only worked with the SQLite online store. All other online stores raised `VersionedOnlineReadNotSupported` when a versioned ref was used.

This PR adds versioned read/write support to the FAISS online store, following the same pattern as the SQLite reference implementation:

- **Write path**: when `enable_online_feature_view_versioning` is enabled, `_table_id` appends a `_v{N}` suffix to the in-memory key namespace (e.g. `project_driver_stats_v2`), routing materialization to the correct versioned index
- **Read path**: version-qualified lookups (`@v2`) set `projection.version_tag` on the feature view before `online_read` is called; `_table_id` picks this up and routes to the correct index. `projection.version_tag` takes priority over `current_version_number` so explicit version requests are always honoured
- **Multi-table support**: replaced single class-level `_index`/`_in_memory_store` with per-table `_indices`/`_in_memory_stores` dicts, so multiple feature views and versions can coexist in memory simultaneously
- **Gate**: `_check_versioned_read_support` in `online_store.py` now allows `FaissOnlineStore` through

All existing `online_read`, `online_write_batch`, `retrieve_online_documents`, `update`, and `teardown` methods are covered since every method resolves its namespace through `_table_id`.

# Which issue(s) this PR fixes:
Part of #2728

# Checks
- [x] I've made sure the tests are passing.
- [x] My commits are signed off (`git commit -s`)
- [x] My PR title follows [conventional commits](https://www.conventionalcommits.org/) format

## Testing Strategy
- [x] Unit tests
- [ ] Integration tests
- [ ] Manual tests
- [ ] Testing is not required for this change

Added 11 unit tests covering `_table_id` with versioning disabled, `projection.version_tag` priority over `current_version_number`, version 0 edge case, no version info fallback, versioned write/read round trips, namespace isolation between versions, missing index handling, and teardown cleanup.

# Misc
- Removed `"type"` from mock config in tests to match `FaissOnlineStoreConfig`'s strict pydantic validation (no extra fields permitted)

---

<a href="https://app.devin.ai/review/feast-dev/feast/pull/6196" target="_blank">
  <picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://static.devin.ai/assets/gh-open-in-devin-review-dark.svg?v=1">
    Open with Devin
  </picture>
</a>

### CODE REVIEW

File: sdk/python/feast/infra/online_stores/online_store.py [hunk 1]
```diff
@@ -256,9 +256,10 @@ def get_online_features(
 
     def _check_versioned_read_support(self, grouped_refs):
         """Raise an error if versioned reads are attempted on unsupported stores."""
+        from feast.infra.online_stores.faiss_online_store import FaissOnlineStore
         from feast.infra.online_stores.sqlite import SqliteOnlineStore
 
-        if isinstance(self, SqliteOnlineStore):
+        if isinstance(self, (SqliteOnlineStore, FaissOnlineStore)):
```
[the remark this answers is not in the document]
Comment (ntkathole) [reply sdk/python/feast/infra/online_stores/online_store.py]: @rpathade this is critical

File: sdk/python/feast/infra/online_stores/faiss_online_store.py [hunk 2]
```diff
@@ -43,16 +43,30 @@ def teardown(self):
...
+        if version is not None and version > 0:
+            name = f"{table.name}_v{version}"
+    return f"{project}_{name}"
+
+
 class FaissOnlineStore(OnlineStore):
-    _index: Optional[faiss.IndexIVFFlat] = None
-    _in_memory_store: InMemoryStore = InMemoryStore()
+    _indices: Dict[str, faiss.IndexIVFFlat] = {}
+    _in_memory_stores: Dict[str, InMemoryStore] = {}
```
Comment (ntkathole) [inline sdk/python/feast/infra/online_stores/faiss_online_store.py]: may be good to initialize these in `__init__` ? may cause issue with tests creating multiple objects of class

**Comment by ntkathole** [comment]
This is fixed via https://github.com/feast-dev/feast/pull/6256, Thanks @rpathade
